### PR TITLE
rethinkdb: add merge & coerceTo, improve getAll signature with overloads

### PR DESCRIPTION
**Pull request by gristow**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: <<https://www.rethinkdb.com/api/javascript/coerce_to/>><<https://www.rethinkdb.com/api/javascript/merge/>>
- [x] Increase the version number in the header if appropriate. (Not appropriate -- did not increase)
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

### CODE REVIEW

File: types/rethinkdb/index.d.ts [hunk 1]
```diff
@@ -270,13 +270,28 @@ declare module "rethinkdb" {
         get<TObjectType extends object>(key: string): Operation<TObjectType | null> & Writeable;
         getAll(key: string, index?: Index): Sequence; // without index defaults to primary key
         getAll(...keys: string[]): Sequence;
+        getAll(...keys: string[]): Sequence;
```
Comment (Hoishin) [inline types/rethinkdb/index.d.ts]: Duplicate lines.
Also I think each overload needs `index`.

**Review by Hoishin — CHANGES_REQUESTED** [review]

File: types/rethinkdb/index.d.ts (hunk 1 above)
Comment (Hoishin) [reply types/rethinkdb/index.d.ts]: nvm about `index`. TypeScript doesn't support additional parameters after rest parameters...

**Comment by gristow** [comment]
`.getAll` is just massively overloaded as a function, so to catch all of it takes a lot. I've done my best.

Sorry about the duplicate lines in the prior commit -- I got rid of those.

I also added support for compound indexes which hadn't been there before.

To deal with the impossible `getall(...keys: string[], index?: index` signature, I added overloads for up to five key properties (both for simple & compound indexes).

Let me know if you see any other errors!

File: types/rethinkdb/index.d.ts [hunk 2]
```diff
@@ -268,15 +268,124 @@ declare module "rethinkdb" {
...
+            key4: string,
+            key5: string,
+            index?: Index
+        ): Sequence; // without index defaults to primary key
+        /**
+         * Get all documents matching a compound index key.
+         * See [getAll](https://www.rethinkdb.com/api/javascript/get_all/) and
+         * [compound indexes](https://www.rethinkdb.com/docs/secondary-indexes/javascript/#compound-indexes)
+         */
+        getAll(compoundKey: string[], index: Index): Sequence; // without index defaults to primary key
```
Comment (Hoishin) [inline types/rethinkdb/index.d.ts]: One of overloads with `compoundKey` have `index` optional. Did you mean to have the one required or have all the overloads optional?

**Review by Hoishin — CHANGES_REQUESTED** [review]

**Comment by gristow** [comment]
Thanks for catching -- none optional for compound indexes. Fixed.

**Review by Hoishin — APPROVED** [review]

**Comment by Hoishin** [comment]
Thank you!

**Comment by uniqueiniquity** [comment]
Congratulations on your first DefinitelyTyped contribution!
Thank you for being a part of the community!

**Pull request merged**